### PR TITLE
Refactor networking/serialization

### PR DESCRIPTION
**Pull request by dustfire**
This refactor makes underlying networking classes public and `open` for inheritance for further extension of the SDK. 
We also remove the `GPHRequestType` and dependency to it everywhere.  
Modifying `GPHMappable` protocol so its not hardcoded to a certain type, enabling passing `options` instead of function arguments which were fully hardcoded before. 
We also modified the public Client interface and made it more verbose to see which end-points and what parameters are sent. 
This refactoring can also go further and consolidate AbstractClient further by removing common functions and but for now it is cleaner and more extensible than the older version(s) of the SDK by allowing 3rd party developers to extend Client and use underlying networking layer.

### CODE REVIEW

File: Source/GPHClient.swift [hunk 1]
```diff
@@ -181,9 +201,13 @@ public typealias GPHCompletionHandler<T> = (_ response: T?, _ error: Error?) ->
     @discardableResult public func gifsByIDs(_ ids: [String],
                                      completionHandler: @escaping GPHCompletionHandler<GPHListMediaResponse>) -> Operation {
     
-        let request = GPHRequestRouter.getAll(ids).asURLRequest(apiKey)
-        
-        return self.listRequest(with: request, type: .getAll, media: .gif, completionHandler: completionHandler)
+
+        // Build the request endpoint
+        let queryItems:[URLQueryItem] = [
+            URLQueryItem(name: "ids", value: ids.flatMap({$0}).joined(separator:","))
```
Comment (genegoykhman) [inline Source/GPHClient.swift]: Do you need the flatMap here or can you just do `ids.joined(separator: ",")`?

File: Source/GPHClient.swift [hunk 2]
```diff
@@ -269,9 +310,8 @@ public typealias GPHCompletionHandler<T> = (_ response: T?, _ error: Error?) ->
     @discardableResult public func termSuggestions(_ term: String,
                                                    completionHandler: @escaping GPHCompletionHandler<GPHListTermSuggestionResponse>) -> Operation {
         
-        let request = GPHRequestRouter.termSuggestions(term).asURLRequest(apiKey)
-        
-        return self.listTermSuggestionsRequest(with: request, type: .termSuggestions, media: .gif, completionHandler: completionHandler)
+        let request = GPHRequestRouter.request("queries/suggest/\(term)", .get, nil).asURLRequest(apiKey)
```
Comment (genegoykhman) [inline Source/GPHClient.swift]: This gets escaped at some point right?

File: Source/GPHRequest.swift [hunk 3]
```diff
@@ -144,62 +95,59 @@ class GPHRequest: GPHAsyncOperationWithCompletion {
...
+
+/// Router to generate URLRequest objects.
+///
+public enum GPHRequestRouter {
+    /// MARK: Properties
     
-    /// Get channel gifs+stickers endpoint: id, offset, limit
-    case channelContent(Int, Int, Int)
+    /// Setup the Request: Path, Method, Parameters)
+    case request(String, GPHRequestType, [URLQueryItem]?)
```
Comment (genegoykhman) [inline Source/GPHRequest.swift]: If this is the only possibility, does `GPHRequestRouter` still need to be an enum?

File: Source/GPHClient.swift (hunk 1 above)
Comment (dustfire) [reply Source/GPHClient.swift]: hmmmm good catch... tests should fail too.

File: Source/GPHClient.swift (hunk 2 above)
Comment (dustfire) [reply Source/GPHClient.swift]: another good catch, yeah need to escape. 👍

File: Source/GPHClient.swift (hunk 1 above)
Comment (dustfire) [reply Source/GPHClient.swift]: well not really.. now I remember why I did this. so if someone sends an array like [ ["edsf"], "sdfsdfsd, "sdfsf" ] it flattens it first.. then joins.. its an extra caution step. but again redundant. removed!

**Pull request merged**